core/unit_test/TestPolicyConstruction.hpp

```diff
@@ -732,4 +732,44 @@ TEST(TEST_CATEGORY_DEATH, policy_bounds_unsafe_narrowing_conversions) {
 }
 #endif
 
+template <class Policy>
+void test_prefer_desired_occupancy(Policy const& police) {
+  static_assert(!Policy::experimental_contains_desired_occupancy, "");
+
+  auto const police_with_occ = Kokkos::Experimental::prefer(
+      police, Kokkos::Experimental::DesiredOccupancy{33});
+  static_assert(
+      decltype(police_with_occ)::experimental_contains_desired_occupancy, "");
+  EXPECT_EQ((int)police_with_occ.experimental_get_desired_occupancy(), 33);
+}
+
+TEST(TEST_CATEGORY, desired_occupancy) {
+  test_prefer_desired_occupancy(Kokkos::RangePolicy<TEST_EXECSPACE>{});
+  // FIXME_SYCL requires MDRange policy and Team policy
+#ifndef KOKKOS_ENABLE_SYCL
+  test_prefer_desired_occupancy(
+      Kokkos::MDRangePolicy<TEST_EXECSPACE, Kokkos::Rank<2>>{});
+  test_prefer_desired_occupancy(Kokkos::TeamPolicy<TEST_EXECSPACE>{});
+#endif
+}
+
+template <class... Args>
+struct DummyPolicy : Kokkos::Impl::PolicyTraits<Args...> {};
+
+TEST(TEST_CATEGORY, desired_occupancy_empty_base_optimization) {
+  DummyPolicy<> const policy_empty{};
+  static_assert(sizeof(decltype(policy_empty)) == 1, "");
+  static_assert(
+      !decltype(policy_empty)::experimental_contains_desired_occupancy, "");
+
+  using Kokkos::Experimental::DesiredOccupancy;
+  DummyPolicy<DesiredOccupancy> const policy_with_desired_occupancy{};
```

Comment:
You should use `prefer` rather than adding the trait by hand, since that's the behavior we want to be stable.

Reply:
Fixed by adding a converting constructor to `PolicyDataStorage` and updating the one in `RangePolicy`.

`MDRangePolicy` and `TeamPolicy` will need to be updated as well.